src/java/main/org/apache/zookeeper/client/StaticHostProvider.java

```diff
@@ -87,15 +75,106 @@ public StaticHostProvider(Collection<InetSocketAddress> serverAddresses)
+     * This method is to provide a replacement for InetSocketAddress.getHostString().
+     *
+     * @param addr
+     * @return Hostname string of address parameter
+     */
+    private String getHostString(InetSocketAddress addr) {
+        String hostString = "";
+
+        if(addr == null) {
+            return hostString;
+        }
+        if (!addr.isUnresolved()) {
+            InetAddress ia = addr.getAddress();
+
+            // If the string starts with '/', then it has no hostname
+            // and we want to avoid the reverse lookup, so we return
+            // the string representation of the address.
+            if (ia.toString().startsWith("/")) {
+                hostString = ia.getHostAddress();
+            } else {
+                hostString = addr.getHostName();
+            }
+        } else {
+            // According to the Java 6 documentation, if the hostname is
+            // unresolved, then the string before the colon is the hostname.
+            String addrString = addr.toString();
+            hostString = addrString.substring(0, addrString.lastIndexOf(':'));
+        }
+
+        return hostString;
+    }
+
     public int size() {
         return serverAddresses.size();
     }
 
+    // Counts the number of addresses added and removed during
+    // the last call to next. Used mainly for test purposes.
+    // See StasticHostProviderTest.
+    private int nextAdded = 0;
+    private int nextRemoved = 0;
+
+    int getNextAdded() {
+        return nextAdded;
+    }
+
+    int getNextRemoved() {
+        return nextRemoved;
+    }
+
     public InetSocketAddress next(long spinDelay) {
-        ++currentIndex;
-        if (currentIndex == serverAddresses.size()) {
-            currentIndex = 0;
+        // Handle possible connection error by re-resolving hostname if possible
+        if (!connectedSinceNext) {
+            InetSocketAddress curAddr = serverAddresses.get(currentIndex);
+            String curHostString = getHostString(curAddr);
+            if (!getHostString(curAddr).equals(curAddr.getAddress().getHostAddress())) {
+                LOG.info("Resolving again hostname: {}", getHostString(curAddr));
+                try {
+                    int thePort = curAddr.getPort();
+                    InetAddress resolvedAddresses[] = InetAddress.getAllByName(getHostString(curAddr));
+                    nextAdded = 0;
+                    nextRemoved = 0;
+                    if (resolvedAddresses.length == 1) {
+                        serverAddresses.set(currentIndex, new InetSocketAddress(resolvedAddresses[0], thePort));
+                        nextAdded = nextRemoved = 1;
+                        LOG.debug("Newly resolved address: {}", resolvedAddresses[0]);
+                    } else {
+                        int i = 0;
+                        while (i < serverAddresses.size()) {
+                            if (getHostString(serverAddresses.get(i)).equals(curHostString) &&
```

Comment:
I think that this fixes the issue I described.

nit: Just wondering if it would be easier to use a map from (hoststring,port) -> serverAddress to make things clearer/remove a loop?